include/taskolib/UniqueId.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * \file   UniqueId.h
+ * \author Lars Fröhlich
+ * \date   Created on July 26, 2023
+ * \brief  Declaration of the UniqueId class.
+ *
+ * \copyright Copyright 2023 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#ifndef TASKOLIB_UNIQUEID_H_
+#define TASKOLIB_UNIQUEID_H_
+
+#include <random>
+#include <string>
+
+#include <gul14/optional.h>
+#include <gul14/string_view.h>
+
+namespace task {
+
+/**
+ * An unsigned 64-bit integer for use as a unique ID.
+ */
+class UniqueId
+{
+public:
+    using ValueType = std::uint64_t;
+
+    /// Default-construct a random unique ID.
+    UniqueId();
+
+    /// Construct a unique ID from a given integer.
+    explicit UniqueId(ValueType id);
+
+    /**
+     * Create a unique ID from the given string, returning an empty optional if the string
+     * does not represent a valid hexadecimal number.
+     *
+     * The string must represent a hexadecimal number with a maximum of 16 characters.
+     * Neither leading whitespace nor a "0x" prefix are allowed.
+     */
+    static gul14::optional<UniqueId> from_string(gul14::string_view str);
+
+    /// Determine if two unique IDs are equal.
+    friend bool operator==(UniqueId a, UniqueId b)
+    {
+        return a.id_ == b.id_;
+    }
+
+    /// Determine if two unique IDs are different.
+    friend bool operator!=(UniqueId a, UniqueId b)
+    {
+        return a.id_ != b.id_;
+    }
+
+    /// Return a hexadecimal string representation of the given unique ID.
+    friend std::string to_string(UniqueId uid);
+
+private:
+    static thread_local std::mt19937_64 random_number_generator_;
+
+    ValueType id_;
+};
+
+
+namespace literals {
+
+/// User-defined literal for generating a UniqueId.
+inline ::task::UniqueId operator""_uid(unsigned long long id)
```

Comment:
Only `unsigned long long` is allowed for an integer user-defined literal. See https://en.cppreference.com/w/cpp/language/user_literal:

> Only the following parameter lists are allowed on literal operators:
>
>( const char* ) | (1) |  
>( unsigned long long int ) | (2) |  
>( long double ) | (3) |  
>( char ) | (4) |  
>( wchar_t ) | (5) |  
>( char8_t ) | (6) | (since C++20)
>( char16_t ) | (7) |  
>( char32_t ) | (8) |  
>( const char*, std::size_t ) | (9) |  
>( const wchar_t*, std::size_t ) | (10) |  
>( const char8_t*, std::size_t ) | (11) | (since C++20)
>( const char16_t*, std::size_t ) | (12) |  
>( const char32_t*, std::size_t ) | (13) |  
>
>(1) Literal operators with this parameter list are the <i>raw literal operators</i>, used as fallbacks for integer and floating-point user-defined literals (see above)
>(2) Literal operators with these parameter lists are the first-choice literal operator for user-defined integer literals
>(3) Literal operators with these parameter lists are the first-choice literal operator for user-defined floating-point literals
>(4-8) Literal operators with these parameter lists are called by user-defined character literals
>(9-13) Literal operators with these parameter lists are called by user-defined string literals